### ISSUE TITLE
Problematic sentences of Central Kurdish

### ISSUE THREAD

**Issue by Sarchia** [common-voice/sentence-collector#568]
Hi,
As you know Central Kurdish recently launched to contribute. Unfortunately, immediately I noticed there are many incorrect or problematic sentences. I don't know who approved them! but it seems they don't know much about the standard orthography of Central Kurdish. Other problems include missing characters and using non-Kurdish Unicode characters.

I reported many sentences! but I thought the better way is to ask you here to delete these sentences from the database, in order not to show them to contributors.

I also submitted several thousand sentences.  If you need me to identify the incorrect sentences, I need you to show me a text file of sentences that were available a week ago.

**Comment by MichaelKohler** [comment]
Hi, thanks for reporting this. Here is the list of all sentences that were part of the last release that happened 2 weeks ago: https://github.com/common-voice/common-voice/blob/68b7b1a1ebcbdeac6f02b7fb8cef141e8240b22c/server/data/ckb/sentence-collector.txt

If you identify any wrong sentence, you can look at its details by using this link and adjusting the sentence at the end:

https://commonvoice.mozilla.org/sentence-collector/sentences/ckb?sentence=%DA%AF%D9%88%D8%B1%D8%B2%DB%8C%E2%80%8C%20%D8%B3%DB%8C%DB%8E%20%D8%B3%DB%95%D8%AA%E2%80%8C%20%D8%A8%D8%A7%D8%AA%D9%85%D8%A7%D9%86%DB%8C

Every sentence has a "source" and a "batch" field. If you notice a pattern there, such as everything coming from the same source or even the same "batch", we can delete based on that. However for both cases we'd need to be a bit careful to not delete a lot of valid sentences with it as well.

If all that fails, we can also delete sentence by sentence if you can give me a text file with a sentence per line that should be deleted.

Then I would also like to hear a bit more about the mistakes. We have a mechanism to add a specific validation rule to uploads to the Sentence Collector: https://github.com/common-voice/sentence-collector/blob/main/server/lib/validation/VALIDATION.md . That might be able to catch these kind of errors for the future.

**Comment by Sarchia** [comment]
I tried your link to know the source of problematic sentences, but I thought it is better to review the list of the sentences. Although it was a headache, I tried to identify all of them. I attached the text file of problematic sentences here which contains 1992 sentences. I noticed most of them are from a source named "folk".

Besides removing these sentences, I don't know what will be your plan for the audio records of these sentences?!
But I think it's better to delete them too.

Regarding creating rules or a filter, I suggest you reject sentences with some Arabic and Persian characters which are not part of Central Kurdish alphabet. Some of them are (ي, ؤ, ة, أ, إ, آ, ك, ى), plus (U+200C) code, which doesn't have a shape but used for splitting letters, many sentences had this last character, so it's necessary to reject sentences if they include it.

[Problematic sentences CKB.txt](https://github.com/common-voice/sentence-collector/files/7644956/Problematic.sentences.CKB.txt)

**Comment by MichaelKohler** [comment]
Thanks for doing this, I will take care of this in the next few days.

Are the 34 letters listed in the "Sorani script" section on https://en.m.wikipedia.org/wiki/Kurdish_alphabets all that should be allowed? Might be easier to only allow certain than blocking some and missing others.

**Comment by Sarchia** [comment]
As people contribute to the language every day, I hope these sentences be deleted soon.

Maybe some issues arise by limiting characters to this list, so I think blocking is a better option here.
Also, I don't know if you have a converter tool, but If this is possible, you can also convert some of those letters instead of rejecting them. For example, ك to ک and ي to ی.

**Comment by MichaelKohler** [comment]
Converting is currently only possible when exporting to Common Voice, however I like the idea if doing that when uploading. I think that could be implemented. Do you have a full list of these conversations?

For blocking vs. allowing: in theory it would be possible to add sentences with English or Russian characters as well as it is right now. All these would currently need to be rejected when reviewing them. Therefore generally an allow-list would be better in my opinion. However that would need to be exhaustive. In either case I'd need the full list to create the rules file.

**Comment by Sarchia** [comment]
I still think blocking these letters can do a good job here, you can add these characters to the list too (ص ,ض, ط, ظ, ذ, ث, ۀ), and plus English letters If you think they should be excluded.

If this is okay, I can explain another way to block some other problematic sentences.

**Comment by MichaelKohler** [comment]
Ok, let's do it this way and see how it works out. Happy to hear your other thoughts and then I will create a rules file.

**Comment by Sarchia** [comment]
In Kurdish (Sorani) orthography, there is a rule which tells these vowels (ێ, ا, ە, ۆ) can't be written after each other. There should be other letters between them. Also, there is no word starting with these letters.

Another rule is that no words start with these two letters (ر) and (ڵ), it should start with (ڕ) for the first one and (ل) for the second one.

I'm not sure if a filter is needed for that, I saw some sentences having some problems, but not much as other problems.

Here is also the Persian numbers (۰ ۱ ۲ ۳ ۴ ۵ ۶ ۷ ۸ ۹), it is better to block them too, they are similar to Kurdish numbers but have different codes. And about English numbers, I'm not sure what to do with them, many people and news agencies use them.

**Comment by Sarchia** [comment]
And if you still believe a list of allowed characters is better, I can give you details about it too.

**Comment by HarikalarKutusu** [comment]
@Sarchia, please check this one:
https://github.com/ftyers/commonvoice-utils/blob/main/cvutils/data/ckb/alphabet.txt

**Comment by Sarchia** [comment]
@HarikalarKutusu  The link contains all Central Kurdish letters, but this (-) is not a letter and some letters were repeated.

**Comment by HarikalarKutusu** [comment]
That particular alphabet I linked is used to filter out sentences containing letters from foreign alphabets before feeding them into training. Thus it contains "allowed" symbols, including "-"...
For example, although we don't have X and W in the Turkish alphabet, w and x are allowed in Turkish sentences.
Otherwise, some proper names like "Washington", or some Kurdish names would be filtered out.
So, not allowing commonly used foreign characters at this stage can be problematic for training/inference. One can already filter out these to get "pure Sorani" at a later stage.
@MichaelKohler, @ftyers what do you think?

**Comment by MichaelKohler** [comment]
I think for now it's fine to go with a blocklist. I'll create a rules file and we can see from there.

**Comment by HarikalarKutusu** [comment]
> Here is also the Persian numbers (۰ ۱ ۲ ۳ ۴ ۵ ۶ ۷ ۸ ۹), it is better to block them too, they are similar to Kurdish numbers but have different codes. And about English numbers, I'm not sure what to do with them, many people and news agencies use them.

Numerals shouldn't be there anyway...

**Comment by MichaelKohler** [comment]
@Sarchia I've created an initial version of the validation file. Whenever a validation fails, we show an error message in the GUI of the Sentence Collector. Right now these are English, but it would be great if you could translate the following sentences to Central Kurdish for me:

* Number of words must be between 1 and 14 (inclusive)
* Sentence should not contain numbers
* Sentence should not contain symbols (translation note: symbols here refers to symbols such as @, #, etc)
* Sentence should not contain non-Central-Kurdish characters

Thanks!

**Comment by Sarchia** [comment]
Thank you @MichaelKohler 
I didn't know numbers are not allowed. There are some approved sentences with numbers. In addition to the file of problematic sentences, you can easily find sentences with numbers to remove them. Here are Kurdish numbers: ٠ ١ ٢ ٣ ٤ ٥ ٦ ٧ ٨ ٩

Also, I want to know if punctuation marks should be allowed? Especially the percentage mark (٪) which usually is written with numbers.


Central Kurdish translations of mentioned sentences:
- ژمارەی وشەکان دەبێت لەنێوان ١ بۆ ١٤ بێت
- نابێت ڕستە ژمارەى تێدا بێت
- نابێت ڕستە هێمای تێدا بێت (وەک @ و #)
- نابێت ڕستە پیتی بەدەر لە ئەلفوبێی کوردی (سۆرانی) تێدا بێت

**Comment by MichaelKohler** [comment]
> Also, I want to know if punctuation marks should be allowed? Especially the percentage mark (٪) which usually is written with numbers.

I don't see the point of allowing %. In general punctuation marks are allowed though. I will filter those out as well.

Thanks for the translations!

**Comment by rojserbest** [comment]
Thanks, @MichaelKohler & @Sarchia!

**Comment by Sarchia** [comment]
@MichaelKohler  I just wanted to know what will happen to audio recordings of the problematic sentences?  They are still showing for review, and I think many of them were accepted by other users.

**Comment by Sarchia** [comment]
Dear @MichaelKohler I saw from [these examples](https://huggingface.co/datasets/mozilla-foundation/common_voice_9_0/viewer/ckb/train) that the problematic sentences and their recordings were published in the 9th version. I'm not sure about the the version 10, if this is the case, please delete the problematic sentences and their recordings in these datasets and the upcoming versions.

**Comment by Heyhillary** [comment]
Hey @Sarchia ,

Following up from the Hugging Face message. 

Is it possible to know if the new examples you are referring to, the issues that they have (if they are different from the previous ones) ?

Thank you !

**Comment by Sarchia** [comment]
Hi @Heyhillary 
The file which I put on Hugging Face, is from this issue. I think these sentences are no longer showing to contributors, but the problem is these problematic sentences and their clips are released with the datasets, as we saw in the data of 9th version. These sentences should be excluded.

Also, I flagged many other sentences whenever I saw a wrong sentence when contributing. If it is possible, I suggest you to show me a list of flagged sentences, then we can fix the minor mistakes or remove them if there are big issues with the sentences.
